### ISSUE TITLE
Change the way we teach AJAX

### ISSUE THREAD

**Issue by lirenyeo** [NextAcademy/ground-control-center#10]
For some unknown reasons we have instilled fears of AJAX among students. No doubt this is a fault we need to correct.

During the taylor's kidcamp, I introduced API first, then only I tell them how AJAX is just a method to make API calls through JS. I think it was effective, I will write out my methods below for reference:

#### 1. What is server and http request
- Show the usual http request diagram
- Think of server as a house with doors, by knowing the right address (IP), we can go knock on the door, and GET or POST (give) stuffs from/to the household.
- The usual way we do it is through **browser** (Post office 📮). We own a post office (Chrome, IE), then we send mailman 👨‍✈️  to visit these houses, knock on the door and GET/POST stuffs. These mailmen will then come back to our office with the packages 📁 (`html`, `css` and `js`).
- Our browsers will understand these files, and display accordingly _(except IE because the workers in IE are too old to understand the new stuffs and they don't want to teach themselves about CSS grid and flexbox, shame on them!! 😡)_
> For example: through browser address bar is `GET`, through sign up button is `POST` (we give them our email, password, etc).

#### 2. What is API
- So actually everything in # 1 is called `Web Service`. There is another way to visit the house, it's called `Web API`.
- If the house (server) owners wish to, they can install windows (API endpoints) to their houses.
- To access the house through windows, you need to send pigeons 🕊️ (write code) to these houses and let the pigeons GET/POST information from it.
- Unlike sending a mailman👨‍✈️, sending pigeons🕊️ usually give you back `JSON` file.
- How do you find the addresses of these windows? You use API documentation
> Show students an example of API documentation (giphy, tumblr, etc)

#### 3. How to call API?
- Depends on what kind of pigeons you send - what kind of language you write!
- In ruby, you can perhaps use [http gem](https://github.com/httprb/http) to get stuffs from those houses (servers).
- If it's the [GIPHY house](https://developers.giphy.com/), then your pigeon can bring you a random GIF from the server.
- How you decide to use the `JSON` files 📁 you retrieved is up to you.
> Make a request to [chuck norris api](https://api.chucknorris.io/jokes/random) and show them how JSON looks like

#### 4. What is AJAX?
- not [this](https://www.google.com/search?q=ajax+fabuloso&source=lnms&tbm=isch&sa=X&ved=0ahUKEwj6lr3QlaDaAhVFtY8KHYHtBf0Q_AUICigB&biw=1440&bih=781)
- If we write javascript to make API calls (send pigeons), and with the package (JSON) we get back, we use javascript to change how our website looks, it's called AJAX!
- **So basically AJAX is just a Javascript method/technique**
- Like how we use `http` ruby gem, we need js libraries that allow us to make API calls easily, for example [axios](https://github.com/axios/axios), [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) or [jquery](http://api.jquery.com/jquery.ajax/).
- With the JSON we retrieved, we do **DOM Manipulation** to change how our website looks, and that's ajax for you!
- SURPRISE! We can also send our pigeons to our internal RESTful routes and get stuffs! (Make them realize the routes are actually APIs if they expose it to outside)
> Give them a small assignment on APIs, I usually ask them to make a [random gif generator](https://gifmesomething.herokuapp.com/).

**Comment by Hinsei** [comment]
sir daddy liren :clap::clap::clap:

**Comment by audreylky** [comment]
@lirenyeo 
We may also need to cater for non-jQuery form of AJAX... or at least show them the jQuery version, and then a non-jQuery version side by side.

I think this looks pretty good!

**Comment by audreylky** [comment]
Was thinking that while everyone will have their own way of delivering the lecture, maybe this one can be like a short video explanation for them to watch before the actual lecture?

**Comment by shanqyeet** [comment]
Agree with @audreylky

**Comment by audreylky** [comment]
@coffeestrike @shanqyeet @qaqaqaqaowh @sophieyang0813 
Can discuss where to put this?

@lirenyeo 
Do the video by next week?

**Comment by sophieyang0813** [comment]
I came across this video on how the internet works for developers. While not AJAX specific, I find this helpful and thought I'd share 😃 1)[ part 1: Overview & Frontend]( https://www.youtube.com/watch?v=e4S8zfLdLgQ )  2)[ part 2: Servers & Scaling](https://www.youtube.com/watch?v=FTAPjr7vgxE )
